cou/apps/factory.py

```diff
@@ -34,48 +32,37 @@ class AppFactory:
     charms: dict[str, type[OpenStackApplication]] = {}
 
     @classmethod
-    def create(
-        cls,
-        name: str,
-        status: ApplicationStatus,
-        config: dict,
-        model: COUModel,
-        charm: str,
-        machines: dict[str, COUMachine],
-    ) -> Optional[OpenStackApplication]:
+    def create(cls, app: COUApplication) -> Optional[OpenStackApplication]:
         """Create the OpenStackApplication or registered subclasses.
 
         Applications Subclasses registered with the "register_application"
         decorator can be instantiated and used with their customized methods.
-        :param name: Name of the application
-        :type name: str
-        :param machines: Machines in the model
-        :type machines: dict[str, COUMachine]
-        :param status: Status of the application
-        :type status: ApplicationStatus
-        :param config: Configuration of the application
-        :type config: dict
-        :param model: COUModel object
-        :type model: COUModel
-        :param charm: Name of the charm
-        :type charm: str
+        :param app: COUApplication
+        :type app: COUApplication
         :return: The OpenStackApplication class or None if not supported.
         :rtype: Optional[OpenStackApplication]
         """
         # pylint: disable=too-many-arguments
-        if is_charm_supported(charm):
-            app_class = cls.charms.get(charm, OpenStackApplication)
+        if is_charm_supported(app.charm):
+            app_class = cls.charms.get(app.charm, OpenStackApplication)
             return app_class(
-                name=name,
-                status=status,
-                config=config,
-                model=model,
-                charm=charm,
-                machines=machines,
+                name=app.name,
```

Comment:
I think you can use:
```python
from dataclasses import asdict
return app_class(**asdict(app))
```

Reply:
From the sync it was said that the problem was because asdict creates a deepcopy and this can be problematic to model. Maybe this can be an alternative for shallow copy

```python
return app_class(**dict((field.name, getattr(obj, field.name)) for field in fields(app)))
```